File: cpp/bc2025/src/philip_06/SoldierState/Retreat.cpp
#include "SoldierState/Retreat.hpp"

#include <algorithm>

#include "RobotPlayer.hpp"
#include "Soldier.hpp"
#include "roles/Mopper.hpp"

namespace philip_06 {
namespace SoldierState {

using roles::Mopper;

void Retreat::run(RobotController rc) {
    MapLocation bestOpt = MapLocation::NONE;
    for (const MapLocation& curr : Soldier::paintTowers) {
        if (rc.canSenseLocation(curr) && rc.senseRobotAtLocation(curr).location.isNull()) {
            Soldier::paintTowers.erase(curr);
            break;
        }
    }
    const MapLocation myLoc = rc.getLocation();
    for (const MapLocation& curr : Soldier::paintTowers) {
        if (bestOpt.isNull() ||
            (myLoc.distanceSquaredTo(curr) < myLoc.distanceSquaredTo(bestOpt) &&
             (!rc.canSenseLocation(bestOpt) ||
              !rc.senseRobotAtLocation(bestOpt).location.isNull()))) {
            bestOpt = curr;
        }
    }
    if (bestOpt.isNull()) {
        bestOpt = MapLocation(RobotPlayer::rng.nextInt(rc.getMapWidth()),
                              RobotPlayer::rng.nextInt(rc.getMapHeight()));
        Mopper::moveTowardsMindfully(rc, bestOpt);
        return;
    }
    int transfer = 100 - rc.getPaint();
    if (rc.canSenseLocation(bestOpt) && !rc.senseRobotAtLocation(bestOpt).location.isNull()) {
        transfer = std::min(transfer, rc.senseRobotAtLocation(bestOpt).getPaintAmount());
        rc.setIndicatorString(std::to_string(transfer));
        if (rc.canTransferPaint(bestOpt, -transfer)) rc.transferPaint(bestOpt, -transfer);
    }
    Mopper::moveTowardsMindfully(rc, bestOpt);
}

}  // namespace SoldierState
}  // namespace philip_06

File: py/bc2025/src/philip_06/SoldierState/Retreat.py
# Port of philip_06/SoldierState/Retreat.java. See py/CONVENTIONS.md.
# Module-form imports for bot-internal deps (cyclic import graph) - names
# resolve at call time.
from api import *
import RobotPlayer
import Soldier
from roles import Mopper


class Retreat:
    @staticmethod
    def run(rc):
        bestOpt = None
        for curr in Soldier.Soldier.paintTowers:
            if rc.canSenseLocation(curr) and rc.senseRobotAtLocation(curr) is None:
                Soldier.Soldier.paintTowers.discard(curr)
                break
        for curr in Soldier.Soldier.paintTowers:
            if bestOpt is None or (rc.getLocation().distanceSquaredTo(curr) < rc.getLocation().distanceSquaredTo(bestOpt) and (not rc.canSenseLocation(bestOpt) or rc.senseRobotAtLocation(bestOpt) is not None)):
                bestOpt = curr
        if bestOpt is None:
            bestOpt = MapLocation(RobotPlayer.RobotPlayer.rng.nextInt(rc.getMapWidth()), RobotPlayer.RobotPlayer.rng.nextInt(rc.getMapHeight()))
            Mopper.Mopper.moveTowardsMindfully(rc, bestOpt)
            return
        transfer = 100 - rc.getPaint()
        if rc.canSenseLocation(bestOpt) and rc.senseRobotAtLocation(bestOpt) is not None:
            transfer = min(transfer, rc.senseRobotAtLocation(bestOpt).getPaintAmount())
            rc.setIndicatorString(str(transfer))
            if rc.canTransferPaint(bestOpt, -transfer):
                rc.transferPaint(bestOpt, -transfer)
        Mopper.Mopper.moveTowardsMindfully(rc, bestOpt)
